File: FlattenBinarytree_to_LinkedList.java
TC: O(n)
SC: O(n)

Runtime: 1 ms, faster than 61.34% of Java online submissions for Flatten Binary Tree to Linked List.
Memory Usage: 35.9 MB, less than 100.00% of Java online submissions for Flatten Binary Tree to Linked List.

Approach: We need preorder traversal. Use ArrayList to save the preorder. Then change left node to null and right node to the next
element in the List for every element in the list.


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
     List<TreeNode> list;
    public void flatten(TreeNode root) {
        if(root == null ) return ;
        list = new ArrayList<>();
        preorder(root);
        
         for (int i=0; i<list.size()-1; i++ ) {
             TreeNode temp = list.get(i);
             temp.left = null;
             temp.right = list.get(i+1);
            }
        list.get(list.size()-1).left = null;
        list.get(list.size()-1).right = null;
    }
    private void preorder(TreeNode root){
        if(root == null ) return ;
        list.add(root);
        preorder(root.left);
        preorder(root.right);
    }
}

File: flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Time Complexity: O(n), number of nodes
# Space Complexity: O(h), height of the tree
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        #base case
        if root==None or (root.right==None and root.left==None):return
        #logic
        if root.left!=None:
            self.flatten(root.left)
            rootright=root.right
            root.right=root.left
            root.left=None
            while root.right!=None:
                root=root.right
            root.right=rootright
        self.flatten(root.right)
        